sns/ut/isal_ub.c

```diff
@@ -0,0 +1,276 @@
+#define MMAX    255
+#define KMAX    255
+
+
+static int ub_init(const char *opts M0_UNUSED)
+{
+    srand(1285360231);
+    return 0;
+}
+
+/*
+ * Generate decode matrix from encode matrix and erasure list
+ *
+ */
+
+static int gf_gen_decode_matrix_simple(uint8_t * encode_matrix,
+                    uint8_t * decode_matrix,
+                    uint8_t * invert_matrix,
+                    uint8_t * temp_matrix,
+                    uint8_t * decode_index, uint8_t * frag_err_list, int nerrs, int k,
+                    int m)
+{
+    int i, j, p, r;
+    int nsrcerrs = 0;
+    uint8_t s, *b = temp_matrix;
+    uint8_t frag_in_err[MMAX];
+
+    memset(frag_in_err, 0, sizeof(frag_in_err));
+
+    // Order the fragments in erasure for easier sorting
+    for (i = 0; i < nerrs; i++) {
+        if (frag_err_list[i] < k)
+            nsrcerrs++;
+        frag_in_err[frag_err_list[i]] = 1;
+    }
+
+    // Construct b (matrix that encoded remaining frags) by removing erased rows
+    for (i = 0, r = 0; i < k; i++, r++) {
+        while (frag_in_err[r])
+            r++;
+        for (j = 0; j < k; j++)
+            b[k * i + j] = encode_matrix[k * r + j];
+        decode_index[i] = r;
+    }
+
+    // Invert matrix to get recovery matrix
+    if (gf_invert_matrix(b, invert_matrix, k) < 0)
+        return -1;
+
+    // Get decode matrix with only wanted recovery rows
+    for (i = 0; i < nerrs; i++) {
+        if (frag_err_list[i] < k)    // A src err
+            for (j = 0; j < k; j++)
+                decode_matrix[k * i + j] =
+                    invert_matrix[k * frag_err_list[i] + j];
+    }
+
+    // For non-src (parity) erasures need to multiply encode matrix * invert
+    for (p = 0; p < nerrs; p++) {
+        if (frag_err_list[p] >= k) {    // A parity err
+            for (i = 0; i < k; i++) {
+                s = 0;
+                for (j = 0; j < k; j++)
+                    s ^= gf_mul(invert_matrix[j * k + i],
+                            encode_matrix[k * frag_err_list[p] + j]);
+                decode_matrix[k * p + i] = s;
+            }
+        }
+    }
+    return 0;
+}
+
+static int ub_test(uint32_t k, uint32_t p, uint32_t len)
+{
+    uint32_t i, j, m;
+    uint32_t nerrs = 0;
+
+    // Fragment buffer pointers
+    uint8_t *frag_ptrs[MMAX];
+    uint8_t *recover_srcs[KMAX];
+    uint8_t *recover_outp[KMAX];
+    uint8_t frag_err_list[MMAX];
+
+    // Coefficient matrices
+    uint8_t *encode_matrix, *decode_matrix;
+    uint8_t *invert_matrix, *temp_matrix;
+    uint8_t *g_tbls;
+    uint8_t decode_index[MMAX];
+
+    m = k + p;
+
+    // Allocate coding matrices
+    encode_matrix = malloc(m * k);
+    decode_matrix = malloc(m * k);
+    invert_matrix = malloc(m * k);
+    temp_matrix = malloc(m * k);
+    g_tbls = malloc(k * p * 32);
+
+    if (encode_matrix == NULL || decode_matrix == NULL
+        || invert_matrix == NULL || temp_matrix == NULL || g_tbls == NULL) {
+        printf("Test failure! Error with malloc\n");
```

Comment:
Here, we need to:
```
m0_free(encode_matrix);
m0_free(decode_matrix);
m0_free(invert_matrix);
m0_free(temp_matrix);
m0_free(g_tbls);
```
Please note, m0_free(NULL) is OK.

Reply:
Done.